src/main/java/org/broadinstitute/hellbender/tools/spark/bwa/BwaSparkEngine.java

```diff
@@ -1,206 +1,113 @@
-        final JavaRDD<String> samLines = align(shortReadPairs);
-        final Broadcast<SAMFileHeader>  readsHeaderBroadcast = ctx.broadcast(readsHeader);
-        return samLines.mapPartitions(lineIterator -> {
-            //Note: The parser is stateful and not thread safe.
-            // So we reuse the parser over the whole partition and call it from one thread only.
-            final SAMLineParser samLineParser = new SAMLineParser(new DefaultSAMRecordFactory(), ValidationStringency.SILENT, readsHeaderBroadcast.getValue(), null, null);
-            final List<GATKRead> reads = new ArrayList<>();
-            while (lineIterator.hasNext()) {
-                final String samLine = lineIterator.next();
-                reads.add(new SAMRecordToGATKReadAdapter(samLineParser.parseLine(samLine)));
-            }
-            return reads.iterator();
-        });
     }
 
-    private JavaRDD<Tuple2<ShortRead, ShortRead>> convertToUnalignedReadPairs(final JavaRDD<GATKRead> unalignedReads) {
-        final JavaRDD<List<GATKRead>> unalignedPairs = unalignedReads.mapPartitions(iter -> Iterators.partition(iter, 2));
+    public SAMFileHeader getHeader() { return broadcastHeader.getValue(); }
 
-        return unalignedPairs.map(p -> {
-            final GATKRead read1 = p.get(0);
-            final GATKRead read2 = p.get(1);
-            final String name1 = read1.getName();
-            final String name2 = read2.getName();
-            final byte[] baseQualities1 = SAMUtils.phredToFastq(read1.getBaseQualities()).getBytes();
-            final byte[] baseQualities2 = SAMUtils.phredToFastq(read2.getBaseQualities()).getBytes();
-            return new Tuple2<>(
-                    new ShortRead(name1, read1.getBases(), baseQualities1),
-                    new ShortRead(name2, read2.getBases(), baseQualities2));
-        });
+    public JavaRDD<GATKRead> align(final JavaRDD<GATKRead> unalignedReads) {
+        final Broadcast<SAMFileHeader> broadcastHeader = this.broadcastHeader;
+        final String indexFileName = this.indexFileName;
+        return unalignedReads.mapPartitions(itr -> new ReadAligner(indexFileName, broadcastHeader.value()).apply(itr));
     }
 
-    private JavaRDD<String> align(final JavaRDD<Tuple2<ShortRead, ShortRead>> shortReadPairs) {
-        return shortReadPairs.mapPartitions(iter -> {
-            BWANativeLibrary.load();
-
-            try {
-                final File localRef = BucketUtils.isHadoopUrl(referenceFileName) ? localizeReferenceAndBwaIndexFiles(referenceFileName) : new File(referenceFileName);
-                final BwaIndex index = new BwaIndex(localRef);
-                final BwaMem mem = new BwaMem(index);
-                return Utils.concatIterators(alignChunks(mem, iter));
-            } catch (final IOException e) {
-                throw new GATKException("Cannot run BWA-MEM", e);
-            }
-        });
+    @Override
+    public void close() {
+        broadcastHeader.destroy();
```

Comment:
I'm under the impression that it is necessary.  This guy thinks so, too:
http://apache-spark-developers-list.1001551.n3.nabble.com/Broadcast-Variable-Life-Cycle-td18683.html
But in any case, it just seems cleaner to tell Spark when you know you're done with it, rather than letting it hang around indefinitely.  Even if the finalizer on the driver side calls destroy, the driver VM isn't the VM that's going to be starved for memory and running the GC aggressively (typically).  I want to free resources on the executors as soon as I possibly can.
That's my reasoning, anyway.  I may be all wet.